### DISCUSSION TITLE
Define a line of nodesets: gmsh

### DISCUSSION THREAD

**Discussion by burritotukituki** [idaholab/moose#27061] [Q&A Meshing]
Hi,

I'm currently tackling a groundwater flow problem where flux moves from left to right. My aim is to establish hydraulic head boundary conditions by defining the pressure at the bottom left and right nodes. While this method works seamlessly in 2D models utilizing `ExtraNodesetGenerator`, transitioning to a 3D model introduces the need to define a line of nodesets instead of individual nodes.

I employ Gmsh for mesh generation, but I'm encountering difficulties in defining the line of nodesets. Specifically, when attempting to generate a 1D physical group in a 3D mesh using in gmsh

`addPhysicalGroup(1, [1], name="left_BC")`

I encounter an error upon attempting to import the mesh into MOOSE:

`*** ERROR ***
Invalid dimension dim = 4
[0] ../src/fe/fe_base.C, line 405, compiled Mar  3 2023 at 07:21:18`

Any idea what it going on?

**Comment by GiudGiud** [comment]
Hello

@jwpeterson seems you wrote a lot of the GMSH reader. Do you know how nodesets should be created in Gmsh to be loaded into libMesh as nodesets?

@burritotukituki if the shape of the nodeset is simple enough, you can define it in MOOSE instead with the mesh generators
https://mooseframework.inl.gov/syntax/Mesh/index.html

Guillaume

Guillaume

**Reply by burritotukituki** [reply]
This kind of model typically have very complex natural boundaries... Thanks!!

**Reply by jwpeterson** [reply]
Hello @GiudGiud, 

Sorry for the slow reply, I was out of the office last week. I looked at the gmsh_io.C code, and I found that nodes are being added to nodesets for any 0-dimensional elements that are present in the mesh. For Gmsh files with version < 4.0, for example, we do the following:

```
                // Handle 0-dimensional elements (points) by adding
                // them to the BoundaryInfo object with
                // boundary_id == physical.
                else
                {
                  // This seems like it should always be the same
                  // number as the 'id' we already read in on this
                  // line.  At least it was in the example gmsh
                  // file I had...
                  in >> node_id;
                  mesh.get_boundary_info().add_node
                    (nodetrans[node_id],
                     static_cast<boundary_id_type>(physical));
                }
```

It also appears that we handle setting nodeset names based on the information in the PhysicalNames section.
```
              // If it's zero-dimensional then it's a nodeset
              else if (phys_dim == 0)
                mesh.get_boundary_info().nodeset_name(cast_int<boundary_id_type>(phys_id)) = phys_name;
```

So I think that if @burritotukituki adds the nodes that he wants to be in a nodeset as zero-dimensional "elements" in the Gmsh file, then he might have more luck? I would also suggest that they provide a small version of the .msh file that is not working for them so that we can investigate further.

**Comment by 1runer** [comment]
@burritotukituki 
Did you check the .msh file you generated?

You should have something like the following in the very first lines:
The first column is the dimension, second the nodeset number, third the physical name given
```
$PhysicalNames
8
0 2 "point"
1 3 "line"
2 4 "plane"
2 5 "left"
2 6 "right"
2 7 "bottom"
2 8 "top"
3 1 "block"
$EndPhysicalNames
```

I don't use the python interface and instead the classical .geo file, therefore I cannot help you with the input structure.

**Comment by burritotukituki** [comment]
Hi @1runer and @GiudGiud 
Yes I tried what you suggest, many thanks. I think there is some kind of bug in libMesh, when you import a 3D mesh with a zero dimensional or one dimensional physical property is not being recognized and of course the simulation fails.
Where can I report this bug?

**Reply by 1runer** [reply]
Do you have an MWE with a simple mesh and input-file? I would like to reproduce this. 

The code block in my previous post is from a Gmsh mesh I use currently without problems.
The only problem is that the names get lost during import and you have to call the nodesets by its particular number (e.g. "3" for the line feature).
